titan/model.py

```diff
@@ -358,48 +362,112 @@ def agents_interact(self, rel: "ag.Relationship"):
+                                * loc.params.demographics[race].ppl
+                            )
+                        ):
+                            new_agent = self.pop.create_agent(loc, race, self.time)
+                            if entrance.age_in:
+                                new_agent.age = entrance.age
+                            self.pop.add_agent(new_agent)
+            elif entrance.enter_type == "replace":
+                assert (
+                    self.params.classes.exit[strategy.exit_class].exit_type != "none"
+                ), "Cannot replace without exit"
+                p = entrance.prob
+                for agent in self.exits[strategy.exit_class]:
+                    if self.run_random.random() < p:
+                        new_agent = self.pop.create_agent(
+                            agent.location,
+                            agent.race,
+                            self.time,
+                            agent.sex_type,
+                            agent.drug_type,
+                        )
+                        # age in?
+                        if entrance.age_in:
+                            new_agent.age = entrance.age
+                        # add agent to pop
+                        self.pop.add_agent(new_agent)
+
+    def remove_agent(self, agent):
         """
-        Let agents die and replace the dead agent with a new agent randomly.
+        Remove an agent and the agent's relationships
         """
-        # die stage
-        for agent in self.pop.all_agents:
-            # agent incarcerated, don't evaluate for death
-            if agent.incar.active:
-                continue
-
-            # death rate per 1 person-month
-            p = (
-                prob.get_death_rate(
-                    agent.hiv.active,
-                    agent.hiv.aids,
-                    agent.drug_type,
-                    agent.sex_type,
-                    agent.haart.adherent,
-                    agent.race,
-                    agent.location,
-                    self.params.model.time.steps_per_year,
-                )
-                * self.calibration.mortality
-            )
-
-            if self.run_random.random() < p:
-                self.deaths.append(agent)
-
-                # End all existing relationships
-                for rel in copy(agent.relationships):
-                    rel.progress(force=True)
-                    self.pop.remove_relationship(rel)
-
-        # replace stage
-        for agent in self.deaths:
-            # mark agent component as -1 (no componenet)
-            agent.component = "-1"
-
-            # Remove agent from agent class and sub-sets
-            self.pop.remove_agent(agent)
-
-            new_agent = self.pop.create_agent(
-                agent.location, agent.race, self.time, agent.sex_type, agent.drug_type
-            )
-            self.pop.add_agent(new_agent)
+        for rel in copy(agent.relationships):
```

Comment:
should this be part of the population remove_agent instead? is there ever a case where we'd want to remove an agent, but not end all the relationships?